gtsfm/ui/dot_graph_generator.py

```diff
@@ -0,0 +1,137 @@
+"""
+Creates a DOT graph based on the contents of RegistryHolder.REGISTRY.
+
+Author: Kevin Fu
+"""
+
+from gtsfm.ui.registry import RegistryHolder
+
+from pathlib import Path
+import os
+
+import pydot
+from collections import namedtuple
+
+from typing import Tuple
+from gtsfm.ui.registry import UiMetadata
+
+JS_ROOT = os.path.join(Path(__file__).resolve().parent.parent.parent, "rtf_vis_tool")
+
+
+class DotGraphGenerator:
+    """Generates and saves a graph of all the components in REGISTRY."""
+
+    def __init__(self, test_mode=False):
+        """Create DotGraphGenerator.
+
+        Args:
+            test_mode - boolean flag, only set True when unit testing.
+        """
+
+        # create empty directed graph
+        self._graph = pydot.Dot("my_graph", graph_type="digraph", bgcolor="white")
+        # TODO: fontname seems broken?
+
+        # style constants, see http://www.graphviz.org/documentation/
+        style_consts = {
+            "blue_fillcolor": "lightskyblue",
+            "gray_fillcolor": "gainsboro",
+            "node_style": '"rounded, filled, solid"',
+            "node_shape": "box",
+            "arrow_color": "gray75",
+        }
+        # namedtuple here is faster than default Dict and allows use of dot
+        # operator
+        StyleTuple = namedtuple("StyleTuple", style_consts)
+        self._style = StyleTuple(**style_consts)
+
+        self._test_mode = test_mode
+
+    def _build_graph(self):
+        """Build graph based on RegistryHolder's REGISTRY."""
+
+        # TODO: remove this
+        print("!\n" * 100)
+        print(RegistryHolder.get_registry())
+
+        seen_metadata = set()
+        for cls_name, cls_type in RegistryHolder.get_registry().items():
+            # don't add the base class to the graph
+            if cls_name == "GTSFMProcess":
+                continue
+
+            # don't add any test classes to the graph, unless in testing mode
+            if not self._test_mode and cls_name.startswith("Fake"):
+                continue
+
+            # get UI metadata of class
+            metadata = cls_type.get_ui_metadata()
+
+            # skip duplicates
+            # happens when concrete classes implement an abstract class without overwriting get_ui_metadata()
+            if metadata in seen_metadata:
+                continue
+
+            seen_metadata.add(metadata)
+
+            self._add_metadata_to_graph(metadata)
+
+    def _add_metadata_to_graph(self, metadata: UiMetadata):
+        """Given the UiMetadata from a GTSFMProcess, clean it up, then add to graph."""
```

Comment:
Could we rephrase the language about "clean it up" to be a bit more specific? Thanks!